controllers/workflow_controller.go

```diff
@@ -0,0 +1,454 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/go-logr/logr"
+	addonapiv1 "github.com/keikoproj/addon-manager/api/addon"
+	addonv1 "github.com/keikoproj/addon-manager/api/addon/v1alpha1"
+	pkgaddon "github.com/keikoproj/addon-manager/pkg/addon"
+	"github.com/keikoproj/addon-manager/pkg/common"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	wfcontroller = "workflow-controller"
+)
+
+type wfreconcile struct {
+	client       client.Client
+	log          logr.Logger
+	versionCache pkgaddon.VersionCacheClient
+}
+
+func NewWFController(mgr manager.Manager, stopChan <-chan struct{}, addonversioncache pkgaddon.VersionCacheClient) (controller.Controller, error) {
+	r := &wfreconcile{
+		client:       mgr.GetClient(),
+		log:          ctrl.Log.WithName(wfcontroller),
+		versionCache: addonversioncache,
+	}
+
+	c, err := controller.New(wfcontroller, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.Watch(&source.Kind{Type: &wfv1.Workflow{}}, r.enqueueRequestForOwner())
```

Comment:
This watches all workflows cluster-wide, we cannot allow that, please make this watch specific namespace.

Reply:
Also, the workflows will not be owned by this controller, they are "owned" by Addon, in that vein the Workflow itself has an OwnerRef for the addon and actually that can be used by the watch enqueue methods to parse the Addon owner

Reply:
Our controller is not allowed to cluster-wide coz it is confined by the role and namespace https://github.com/keikoproj/addon-manager/blob/master/config/rbac/role.yaml#L104-L109 
https://github.com/keikoproj/addon-manager/blob/master/config/rbac/role_binding.yaml#L17-L18

But, you are right, we should retain the ownership to Addon, forget to sync up to this branch, fixing...

Reply:
a1ca1b6fe01e95a219e9608663504d052be3a24b